keps/sig-testing/2420-reducing-kubernetes-build-maintenance/README.md

```diff
@@ -0,0 +1,671 @@
+
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+1. Continue maintaining both build systems
+
+This has major drawbacks:
+
+- This continues to eat developer time without significant return, for the most part releases, the bulk of CI (periodic tests), and contributor development use the make build(s).
+
+- Key components of our development setup do not easily port to bazel, so it will continue to be an "also":
+  - code generators: Kubernetes uses a lot of generated code, while Bazel *can* run code generators just fine without checking in the generated sources, we need to check them in for consumption by non-bazel users (i.e. external projects), and as a result we have never leveraged this. These generators are largely incompatible with bazel "philosophy":
+    - much of the k8s codegen uses fake go build tags, which would require an external process (i.e. gazelle extension) to turn into build files
+    - additionally, some codegen relies on certain dependencies, which the gazelle extension would need to figure out
+    - many of the code generators are optimized to load the entire Go tree, then generate lots of code at once, which is somewhat incompatible with Bazel's approach to working on a package-by-package level. Generating code package-by-package is much slower (due to having to reparse the tree each time).
+  - separate "hack/tools" go module for linters etc: not only do linters not work well (because any source change busts the cache anyhow), but rules_go does not do multiple go modules well. Multiple go modules allows us to isolate development dependencies (like linters) from release binary dependencies, easing dependency management
+
+2. Improve bazel integration and drop the make based build
+
+In addition to the points made in 1.) above as to why this is not particularly viable for some of our existing development patterns:
+- We would need to improve support for CGO or eliminate CGO in Kubernetes, both of which would be somewhat expensive to develop
+  - At minimum kubelet definitely requires CGO for OS integration (e.g. selinux)
+  - CGO pkg-config directives do not work in `rules_go`, requiring brittle work-arounds
+  - Cross compiling with CGO under bazel is tricky, and despite @ixdy getting close at one point, never shipped in Kubernetes, let alone portably, or capable of shipping a full release.
+    - @ixdy suspects this would largely have to be reimplemented now
+- This is unlikely to be popular with contributors
+- Large portions of Kubernetes are intended for dual use internally and as exported go libraries ("staging") to be consumed by other projects, where we'd still need to support use without bazel (i.e. checked-in generated code etc.).
```

Comment:
Isn't this already a solved problem even with the current Bazel build? `go build`ing `kuberntes/kubernetes` is atotally different can of worms from `go build`ing re-exported libraries anyway.

Reply:
> Isn't this already a solved problem even with the current Bazel build? go building kuberntes/kubernetes is atotally different can of worms from go building re-exported libraries anyway.

This means that we don't get to gain from integrating deeply with bazel, because all of this code must already work without it. (so for example code generation must not be bazel style, it needs to be checked in / on disk).

Reply:
I'm pretty sure you could integrate deeply with Bazel here, by continuing to use code-generating rules within `kubernetes/kubernetes`, and then leveraging these same rules to export generated code to `staging/` repositories when publishing.

Reply:
Regardless we must ensure this code builds for all supported targets with just `go`, it is not feasible to have a world where we only qualify with bazel, unless the entire go / Kubernetes ecosystem adopts bazel, which is unrealistic.

Testing these with bazel has lead to surprises in the past, see the linked issue about moving away from it in kubernetes/kubernetes.